### PR TITLE
Cancellation tokens & .NET 6

### PR DESCRIPTION
**Pull request by kiminuo**
I'm aware of the https://github.com/MetacoSA/NBitcoin/pull/1085#issuecomment-1039800049 comment but this PR is supposed to show what can be done and I'm not sure what preprocessor directives should be used for this.

Anyway, if this PR makes sense to you, please tell me how to move it forward.

### CODE REVIEW

**Comment by kiminuo** [comment]
Too bad I don't have the right to run CI:

```
1 workflow awaiting approval
First-time contributors need a maintainer to approve running workflows. [Learn more.](https://docs.github.com/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks)
```

**Comment by kiminuo** [comment]
@NicolasDorier Any opinion on this please?

**Comment by molnard** [comment]
Makes sense - it can improve `UnboservedTaskException` coming from NBitcoin in Wasabi?

**Comment by kiminuo** [comment]
> it can improve `UnboservedTaskException` coming from NBitcoin in Wasabi?

I haven't spot any unobserved exception in this code.

**Comment by molnard** [comment]
> > it can improve `UnboservedTaskException` coming from NBitcoin in Wasabi?
> 
> I haven't spot any unobserved exception in this code.

I meant that when you abandon a task and that throws an exception that will be unobserved.

**Comment by NicolasDorier** [comment]
hey @kiminuo sorry, reviewing now!

**Comment by NicolasDorier** [comment]
I will edit this PR to use right #define, but I think that's a good idea

File: NBitcoin/Utils.cs [hunk 1]
```diff
@@ -125,6 +125,16 @@ public static IHDKey[] Derive(this IHDKey hdkey, KeyPath[] keyPaths)
 		}
 		public static async Task WithCancellation(this Task task, CancellationToken cancellationToken)
 		{
+#if NET6_0_OR_GREATER
+			try
+			{
+				await task.WaitAsync(cancellationToken).ConfigureAwait(false);
+			}
+			catch (TimeoutException e)
```
Comment (NicolasDorier) [inline NBitcoin/Utils.cs:133]: why are you catching this? `WaitAsync` doesn't mention it

**Comment by NicolasDorier** [comment]
@kiminuo can you cherry-pick this commit: https://github.com/MetacoSA/NBitcoin/commit/81aca91c9429e68fa88f6049563b604b8bf929c6 ?

File: NBitcoin/Utils.cs (hunk 1 above)
Comment (kiminuo) [reply NBitcoin/Utils.cs]: You are right. I overlooked this.

**Comment by kiminuo** [comment]
> @kiminuo can you cherry-pick this commit: [81aca91](https://github.com/MetacoSA/NBitcoin/commit/81aca91c9429e68fa88f6049563b604b8bf929c6) ?

Yes, done. Thanks.

File: NBitcoin/Utils/StreamExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+			await stream.WriteAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+#else
+			await stream.WriteAsync(buffer, offset, count).WithCancellation(cancellationToken).ConfigureAwait(false);
+#endif
+		}	
+
+		public static async Task FlushCancellableAsync(this NetworkStream stream, CancellationToken cancellationToken)
+		{
+#if !NO_SOCKETASYNC
+			await stream.Flush(cancellationToken).ConfigureAwait(false);
```
Comment (NicolasDorier) [inline NBitcoin/Utils/StreamExtensions.cs:39]: Ping @kiminuo should be FlushAsync here

Comment (kiminuo) [reply NBitcoin/Utils/StreamExtensions.cs]: Yes, I pushed.

**Comment by NicolasDorier** [comment]
making new version

**Pull request merged**